docs/r3/architecture_principles.md

```diff
@@ -2,4 +2,71 @@
 
 The options and patterns discussed here are bounded by a set of background architectural assumptions and aspirations. We are concerned not only with the immediate functional realisability of interop use cases, but also with fostering an interop ecosystem that will remain viable in the longer term as it evolves and expands.
 
-*TODO: list out architecture principles*
+We identify five architectural principles:
+
+1. **Respect finality**. No network should be required to repudiate a transaction finalised on that network because of the failure of a cross-network workflow.
+2. **Avoid nondeterminism**. The success or failure of a cross-network workflow should depend entirely on the fully-orderable sequence of actions carried out within that workflow, and not on observable nondeterminism caused by temporal constraints (time locks).
```

Comment:
Its an interesting perspective differential between a timelock referring to the depth of a block vs an actual enforceable time window, I wonder if there is a linguistically meaningful way of expressing that

Reply:
I think a timelock based on block depth is observably nondeterministic from the point of view of an external system submitting a transaction, which could get into a race with block proposal/finalisation

Reply:
I think the basic idea is that it lets you right logic of the form "this asset it locked. let us say that the block in which the lock occurred (knowable after the fact) was n. If Bob presents the preimage to the following hash (say - condition could be something else) in a transaction that is processed in any block m (m<n+10) then Bob can take the asset.  If no such transaction appears in a block before block n+10 then Alice is permitted to submit a transaction that reverts it to her.

I'm probably not expressing it well but the idea is that the creation of a block at some point imposes constraints on what types of subsequent transactions are valid - and in which blocks they must appear (before/after) in order to be valid.

So you don't know beforehand if your tx will be finalised or in which block it will happen. but you are guaranteed that if Bob has not taken the asset by block n+10, he never can.

Reply:
The determinism I want is that the overall state of the exchange moves forwards in a predictable way depending on the serialised sequence of actions taken by the participants. It's fine if I race with you to make the next move and one of us wins and the other loses. It's not so good if I'm racing with the entire remote network to get my move in before a block number limit expires (which could happen due to the behaviour of any number of otherwise uninvolved parties on the network). I don't see that that's morally any different from "I succeed or fail depending on the system clock time on the server that happens to process the request".

Reply:
Net-net... I agree with the principle of avoiding non-determinism... I'm just not convinced 'I could lose a race, yet the underlying network(s) remain entirely consistent' is a good example to use in the definition!